src/smt/illegal_checker.cpp

```diff
@@ -0,0 +1,200 @@
+    d_illegalKinds.insert(Kind::ARCTANGENT);
+    d_illegalKinds.insert(Kind::ARCCOSECANT);
+    d_illegalKinds.insert(Kind::ARCSECANT);
+    d_illegalKinds.insert(Kind::ARCCOTANGENT);
+    d_illegalKinds.insert(Kind::SQRT);
+    d_illegalKinds.insert(Kind::IAND);
+    d_illegalKinds.insert(Kind::POW2);
+  }
+  if (logicInfo().isTheoryEnabled(theory::THEORY_DATATYPES)
+      && !options().datatypes.datatypesExp)
+  {
+    d_illegalKinds.insert(Kind::MATCH);
+  }
+  if (logicInfo().hasCardinalityConstraints() && !options().uf.ufCardExp)
+  {
+    d_illegalKinds.insert(Kind::CARDINALITY_CONSTRAINT);
+  }
+  if (logicInfo().isTheoryEnabled(theory::THEORY_SETS))
+  {
+    if (!options().sets.setsCardExp)
+    {
+      d_illegalKinds.insert(Kind::SET_CARD);
+    }
+    if (!options().sets.relsExp)
+    {
+      d_illegalKinds.insert(Kind::RELATION_TABLE_JOIN);
+      d_illegalKinds.insert(Kind::RELATION_TRANSPOSE);
+      d_illegalKinds.insert(Kind::RELATION_PRODUCT);
+      d_illegalKinds.insert(Kind::RELATION_JOIN);
+      d_illegalKinds.insert(Kind::RELATION_TCLOSURE);
+      d_illegalKinds.insert(Kind::RELATION_IDEN);
+      d_illegalKinds.insert(Kind::RELATION_JOIN_IMAGE);
+      d_illegalKinds.insert(Kind::RELATION_GROUP);
+      d_illegalKinds.insert(Kind::RELATION_AGGREGATE);
+      d_illegalKinds.insert(Kind::RELATION_PROJECT);
+    }
+  }
+  // unsupported theories disables all kinds belonging to the theory
+  std::unordered_set<theory::TheoryId> unsupportedTheories;
+  if (logicInfo().isTheoryEnabled(theory::THEORY_FP)
+      && !options().fp.fp)
+  {
+    unsupportedTheories.insert(theory::TheoryId::THEORY_FP);
+  }
+  if (logicInfo().isTheoryEnabled(theory::THEORY_FF)
+      && !options().ff.ff)
+  {
+    unsupportedTheories.insert(theory::TheoryId::THEORY_FF);
+  }
+  if (logicInfo().isTheoryEnabled(theory::THEORY_BAGS)
+      && !options().bags.bags)
+  {
+    unsupportedTheories.insert(theory::TheoryId::THEORY_BAGS);
+  }
+  if (logicInfo().isTheoryEnabled(theory::THEORY_SEP)
+      && !options().sep.sep)
+  {
+    unsupportedTheories.insert(theory::TheoryId::THEORY_SEP);
+  }
+  if (!unsupportedTheories.empty())
+  {
+    for (int32_t i = 0;
+       i< static_cast<int32_t>(Kind::LAST_KIND);
+        ++i)
+    {
+      Kind k = static_cast<Kind>(i);
+      // these two kinds are special cased in kindToTheoryId, skip
+      if (k==Kind::UNDEFINED_KIND || k==Kind::NULL_EXPR)
+      {
+          continue;
+      }
+      Trace("ajr-temp") << "Check kind " << k << std::endl;
+      theory::TheoryId tid = theory::kindToTheoryId(k);
+      if (unsupportedTheories.find(tid)!=unsupportedTheories.end())
+      {
+          d_illegalKinds.insert(k);
+      }
+    }
+  }
+}
+
+void IllegalChecker::checkAssertions(Assertions& as)
+{
+  // check illegal kinds here
+  const context::CDList<Node>& assertions = as.getAssertionList();
+  size_t asize = assertions.size();
+  size_t i = d_assertionIndex.get();
+  std::unordered_set<TNode> visited;
```

Comment:
Couldn't this cache be a member of the class in order to avoid traversing the same Node multiple times for multiple checkAssertions calls?

Reply:
This is a tradeoff: while it would be less traversals, that would make it so that assertions are never garbage collected.

Reply:
Note that this method is only checking *new* assertions each step; another option would be to make the visited cache user-context-dependent, what do you think?

Reply:
Yes, I think a user-context-dependent cache would be a good idea here.